dango/dex/src/execute.rs

```diff
@@ -573,6 +583,51 @@ fn clear_orders_of_pair(
                 &order,
             )?;
         }
+
+        // Calculate the volume in USD for the filled order
+        let base_asset_price = querier.query_price(oracle, &base_denom, None)?;
+        let volume = base_asset_price.value_of_unit_amount(filled)?.into_int(); // TODO: Better to store as Decimal?
+
+        // Get the previous volume for the user's address
+        let previous_volume = VOLUMES
+            .prefix(&order.user)
+            .values(storage, None, None, IterationOrder::Descending)
+            .next()
+            .transpose()?
+            .unwrap_or(Uint128::ZERO);
+
+        // Record trading volume for the user's address
+        VOLUMES.save(
+            storage,
+            (&order.user, current_time),
+            &previous_volume.checked_add(volume)?,
+        )?;
```

Comment:
1. If a user has multiple orders in the same block, this will write to the storage multiple times. Better if we collect the updated volumes in a `HashMap` and write once.

Btw, do you have tests for this scenario (multiple orders from the same user in multiple pairs, in the same block)?

2. We should purge volumes data that are too old. We can add a `max_volume_age` parameter (e.g. 3 weeks) to the DEX contract's config, and use `Map::clear` to remove volume data of which `timestamp + max_volume_age` is smaller than `ctx.block.timestamp`.

Reply:
@apollo-sturdy For (1), see: 9af8bb8471d6f2ee8e3c8df262d5152490f4d924

For (2), let's do this in a separate PR.

Reply:
No I don't have any tests for that. Can add it in a separate PR!

Reply:
Yeah please

Reply:
Added here: https://github.com/left-curve/left-curve/pull/649